### ISSUE TITLE
Consumer dtor segfaults if close() has not been called

### ISSUE THREAD

**Issue by jvansanten** [confluentinc/confluent-kafka-python#358]
Description
===========

Consumer's destructor causes a segmentation fault if the object is destroyed after `poll()` has been called at least once, but before `close()` has been called. This brings down the entire Python interpreter. The destructor should probably call `close()` itself.

How to reproduce
================

```
import confluent_kafka
import uuid

config = {
	"bootstrap.servers": "localhost:9092",
	"default.topic.config": {"auto.offset.reset": "smallest"},
	"group.id" : uuid.uuid1(),
}

def crashy():
	consumer = confluent_kafka.Consumer(**config)
	consumer.subscribe(['^.*'])

	consumer.poll()
	
	# this prevents the segfault
	# consumer.close()

crashy()
```

Results in

```
(ampel) [wgs04] ~/scratch/projects/ztf/ampel/Ampel % gdb --args python crashy.py
...
Program received signal SIGSEGV, Segmentation fault.
0x00007fffeffbb089 in CallState_get (h=h@entry=0x7ffff7e20978)
   from /afs/ifh.de/group/amanda/scratch/jvsanten/software/miniconda3/envs/ampel/lib/python3.6/site-packages/confluent_kafka/cimpl.cpython-36m-x86_64-linux-gnu.so
Missing separate debuginfos, use: debuginfo-install glibc-2.17-196.el7.x86_64 libuuid-2.23.2-43.el7.x86_64
(gdb) bt
#0  0x00007fffeffbb089 in CallState_get (h=h@entry=0x7ffff7e20978)
   from /afs/ifh.de/group/amanda/scratch/jvsanten/software/miniconda3/envs/ampel/lib/python3.6/site-packages/confluent_kafka/cimpl.cpython-36m-x86_64-linux-gnu.so
#1  0x00007fffeffbc2e1 in Consumer_rebalance_cb (rk=0x74a9c0, err=RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS,
    c_parts=0x7fffe80065c0, opaque=0x7ffff7e20978)
   from /afs/ifh.de/group/amanda/scratch/jvsanten/software/miniconda3/envs/ampel/lib/python3.6/site-packages/confluent_kafka/cimpl.cpython-36m-x86_64-linux-gnu.so
#2  0x00007fffefd2f4af in rd_kafka_poll_cb (rk=rk@entry=0x74a9c0, rko=rko@entry=0x7fffe8038400,
    cb_type=cb_type@entry=2, opaque=opaque@entry=0x0)
   from /afs/ifh.de/group/amanda/scratch/jvsanten/software/miniconda3/envs/ampel/lib/python3.6/site-packages/confluent_kafka/../../../librdkafka.so.1
#3  0x00007fffefd2f824 in rd_kafka_consumer_close (rk=rk@entry=0x74a9c0)
   from /afs/ifh.de/group/amanda/scratch/jvsanten/software/miniconda3/envs/ampel/lib/python3.6/site-packages/confluent_kafka/../../../librdkafka.so.1
#4  0x00007fffefd31ba6 in rd_kafka_destroy_app (blocking=1, rk=0x74a9c0)
   from /afs/ifh.de/group/amanda/scratch/jvsanten/software/miniconda3/envs/ampel/lib/python3.6/site-packages/confluent_kafka/../../../librdkafka.so.1
#5  rd_kafka_destroy (rk=0x74a9c0) at rdkafka.c:649
#6  0x00007fffeffbd1ca in Consumer_dealloc (self=0x7ffff7e20978)
   from /afs/ifh.de/group/amanda/scratch/jvsanten/software/miniconda3/envs/ampel/lib/python3.6/site-packages/confluent_kafka/cimpl.cpython-36m-x86_64-linux-gnu.so
#7  0x00007ffff7943772 in frame_dealloc (f=0x7ffff0478048) at Objects/frameobject.c:423
#8  0x00007ffff7a04310 in _PyFunction_FastCall (co=<optimized out>, args=<optimized out>, nargs=0,
    globals=<optimized out>) at Python/ceval.c:4909
#9  0x00007ffff7a051f4 in fast_function (kwnames=0x0, nargs=<optimized out>, stack=<optimized out>,
    func=0x7ffff7f4c048) at Python/ceval.c:4941
#10 call_function (pp_stack=pp_stack@entry=0x7fffffffd7e0, oparg=oparg@entry=0, kwnames=kwnames@entry=0x0)
    at Python/ceval.c:4845
#11 0x00007ffff7a099e6 in _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at Python/ceval.c:3322
#12 0x00007ffff7a04c3e in _PyEval_EvalCodeWithName (_co=_co@entry=0x7ffff7ec8e40,
    globals=globals@entry=0x7ffff7f4a168, locals=locals@entry=0x7ffff7f4a168, args=args@entry=0x0,
    argcount=argcount@entry=0, kwnames=kwnames@entry=0x0, kwargs=kwargs@entry=0x0, kwcount=kwcount@entry=0,
    kwstep=kwstep@entry=2, defs=defs@entry=0x0, defcount=defcount@entry=0, kwdefs=kwdefs@entry=0x0,
    closure=closure@entry=0x0, name=name@entry=0x0, qualname=qualname@entry=0x0) at Python/ceval.c:4153
#13 0x00007ffff7a0526d in PyEval_EvalCodeEx (_co=_co@entry=0x7ffff7ec8e40,
---Type <return> to continue, or q <return> to quit---
    globals=globals@entry=0x7ffff7f4a168, locals=locals@entry=0x7ffff7f4a168, args=args@entry=0x0,
    argcount=argcount@entry=0, kws=kws@entry=0x0, kwcount=kwcount@entry=0, defs=defs@entry=0x0,
    defcount=defcount@entry=0, kwdefs=kwdefs@entry=0x0, closure=closure@entry=0x0) at Python/ceval.c:4174
#14 0x00007ffff7a052bb in PyEval_EvalCode (co=co@entry=0x7ffff7ec8e40, globals=globals@entry=0x7ffff7f4a168,
    locals=locals@entry=0x7ffff7f4a168) at Python/ceval.c:730
#15 0x00007ffff7a40132 in run_mod (arena=0x7ffff7f66258, flags=0x7fffffffdaf0, locals=0x7ffff7f4a168,
    globals=0x7ffff7f4a168, filename=0x7ffff01d3170, mod=0x644ca8) at Python/pythonrun.c:1025
#16 PyRun_FileExFlags (fp=fp@entry=0x6a98b0, filename_str=filename_str@entry=0x7ffff0474bf0 "crashy.py",
    start=start@entry=257, globals=globals@entry=0x7ffff7f4a168, locals=locals@entry=0x7ffff7f4a168,
    closeit=closeit@entry=1, flags=flags@entry=0x7fffffffdaf0) at Python/pythonrun.c:978
#17 0x00007ffff7a40297 in PyRun_SimpleFileExFlags (fp=fp@entry=0x6a98b0, filename=<optimized out>,
    closeit=closeit@entry=1, flags=flags@entry=0x7fffffffdaf0) at Python/pythonrun.c:420
#18 0x00007ffff7a40733 in PyRun_AnyFileExFlags (fp=fp@entry=0x6a98b0, filename=<optimized out>,
    closeit=closeit@entry=1, flags=flags@entry=0x7fffffffdaf0) at Python/pythonrun.c:81
#19 0x00007ffff7a5c39d in run_file (p_cf=0x7fffffffdaf0, filename=0x603390 L"crashy.py", fp=0x6a98b0)
    at Modules/main.c:338
#20 Py_Main (argc=argc@entry=2, argv=argv@entry=0x602010) at Modules/main.c:809
#21 0x0000000000400bbc in main (argc=2, argv=<optimized out>) at ./Programs/python.c:69
```

Checklist
=========
Please provide the following information:

 - [x] confluent-kafka-python and librdkafka version (`confluent_kafka.version()` and `confluent_kafka.libversion()`): ('0.11.0', 720896), ('0.9.4', 591103)
 - [x] Apache Kafka broker version: 0.10.1.0
 - [x] Client configuration: `{
	"bootstrap.servers": "localhost:9092",
	"default.topic.config": {"auto.offset.reset": "smallest"},
	"group.id" : uuid.uuid1(),
}`
 - [x] Operating system: EL7, Python 3.6.4 from conda-forge
 - [ ] Provide client logs (with `'debug': '..'` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
Thanks for the good bug report!

I think this has been fixed in the latest release, can't repro it on v0.11.4.

**Comment by jvansanten** [comment]
Indeed it appears fixed; I can't reproduce it with v0.11.4 from PyPI either. Sorry for the noise!
